util/merkle/src/node.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright 2018 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use elastic_array::ElasticArray1024;
+use rlp::*;
+use hashdb::DBValue;
+use primitives::H256;
+
+use nibbleslice::NibbleSlice;
+
+
+#[derive(Eq, PartialEq, Debug)]
+pub enum Node<'a> {
+    Leaf(NibbleSlice<'a>, DBValue),
+
+    Branch(NibbleSlice<'a>, [Option<H256>; 16]),
+}
+
+impl<'a> Node<'a> {
+    /// Decode the `node_rlp` and return the Node.
+    pub fn decoded(node_rlp: &'a [u8]) -> Option<Self> {
+        let r = Rlp::new(node_rlp);
+        match r.prototype() {
+            // Empty node
+            Prototype::Data(0) => return None,
+            // leaf node - first is nibble and second is value
+            Prototype::List(2) => {
+                let slice = NibbleSlice::from_encoded(r.at(0).data());
+
+                return Some(Node::Leaf(slice, DBValue::from_slice(&r.at(1).data())))
+            }
+            // branch node - first is nibbles (or empty), the rest 16 are nodes.
+            Prototype::List(17) => {
+                let mut nodes = [None; 16];
+                for i in 0..16 {
+                    nodes[i] = if r.at(i + 1).is_empty() {
+                        None
+                    } else {
+                        Some(r.at(i + 1).as_val::<H256>())
+                    }
+                }
+
+                return Some(Node::Branch(NibbleSlice::from_encoded(r.at(0).data()), nodes))
+            }
+
+            // something went wrong.
+            _ => panic!("Rlp data is not valid."),
+        }
+    }
+
+    /// Encode the node into RLP.
+    pub fn encoded(node: Option<Self>, size: Option<usize>) -> ElasticArray1024<u8> {
+        match node {
+            Some(Node::Leaf(slice, value)) => {
+                let mut stream = RlpStream::new_list(2);
+                stream.append(&&*slice.encoded());
+                stream.append(&&*value);
```

Comment:
Could you explain these `&&*`s?

Reply:
stream.append get argument as trait, encoding and encoding implement on &[u8]. It is trick to take the [u8] array out of ElasticArray. If we use only one ampersand we'll get ref of ElasticArray

Reply:
`ElasticArray<T>` implements `AsRef<[T]>`.
Can you check again if single ampersand really doesn't work?

Reply:
No. It can't. Unlike `Deref` `AsRef` doesn't work implicitly.
You should call it `&&*s` or `&s.as_ref()`

Reply:
What a shame :(